docs/papers/llm-load-balancing-at-scale-chwbl.md

```diff
@@ -0,0 +1,120 @@
+# LLM Load Balancing at Scale: Consistent Hashing with Bounded Loads
+
+Feb 26, 2025
+
+Posted by [Nick Stogner](https://www.linkedin.com/in/nstogner/), Co-Founder of [Substratus.AI](https://www.substratus.ai/)
+
+## Abstract:
+
+When deploying large language models (LLMs) at scale efficient load balancing is crucial for minimizing latency and maximizing throughput. We explore the application of the Consistent Hashing with Bounded Loads (CHWBL) algorithm in the domain of LLM serving. Our approach dramatically reduces Time To First Token (TTFT) while improving overall Tokens Per Second (TPS) compared to the baseline strategies.
+
+## 1. Introduction
+
+Before an inference engine such as vLLM can begin generating output tokens, it must first process the input prompt during the “prefill phase”. The resulting intermediate representation is stored in a key-value (KV) cache. This cache can be reused across multiple requests that share the same prefix - a feature known as "prefix caching". The impact of prefix caching can be significant in several real world scenarios:
+
+**Multi-turn Conversations:** In applications like chatbots (e.g., ChatGPT) or autonomous AI agents, every new turn appends to an evolving shared context.
+
+<img src="/diagrams/multi-turn-clients.excalidraw.png" style="max-height:700px" alt="Multi-turn conversation diagram">
+
+**Multi-threaded Requests with Shared Context:**
+Scenarios that involve multiple queries against a single long document are particularly sensitive to the efficiency of prefix caching. When requests are executed concurrently, even slight improvements in cache hit rates can lead to substantial reductions in end-to-end latency.
+
+<img src="/diagrams/multi-threaded-shared-context.excalidraw.png" style="max-height:700px" alt="Multi-threaded shared context diagram">
+
+The conventional random routing strategy provided by Kubernetes often results in suboptimal cache utilization, leading to frequent cache evictions and degraded performance. A more consistent routing methodology is would be beneficial to keep a relevant cache on each replica.
```

Comment:
this sentence doesn't make sense: "A more consistent routing methodology is would be beneficial to keep a relevant cache on each replica."

maybe rephrase to:
"A prefix based routing methodology is beneficial to keep a relevant cache on each replica and improve cache hit ratios."